### PR TITLE
responsive images / data purification (pt 1)

### PR DESCRIPTION
**Pull request by kevin-roark**
@r1b big PR but most if it is data changes.

1. Adds a `server` directory that contains modules for the `webhook` (which will be finished in next PR). Broadly, the server:
    1. Pulls remote changes.
    2. Sanitizes / validates slugs
    3. Moves and processes images.
    4. Pushes changes if necessary.
 
2. Step `iii` above adds new smaller sizes of each image, so there is also client code here to make use of that with `srcSet` and `sizes` attributes added to `img` elements throughout the site. This means we should be loading smaller versions of the images when they are thumbnails, and big ones when necessary.

### CODE REVIEW

File: server/git.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  pullRepo,
+  pushChanges,
+}
+
+const gitDir = path.join(__dirname, '../')
+
+function pullRepo () {
+  return new Promise((resolve, reject) => {
+    const repo = simpleGit(gitDir)
+    repo.pull('origin', 'master', err => {
```
Comment (r1b) [inline server/git.js]: Consider `util.promisify`

Comment (r1b) [reply server/git.js]: Same for rest of this file

File: server/images.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+  // update yaml files of all designers
+  await Promise.all(designers.map((designer, index) =>
+    updateDesigner(designer, processedWorks[index])
+  ))
+}
+
+async function processDesignerWorks (data) {
+  const { works } = data
+  return Promise.all(works.map(async (work) => {
+    const imagesToProcess = work.images.filter(image =>
```
Comment (r1b) [inline server/images.js]: Be defensive and check images exists

File: server/images.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    ))
+
+    // create necessary sizes
+    const processedImages = await Promise.all(jpegImages.map(jpegFilename =>
+      resizeImage(jpegFilename)
+    ))
+
+    // return relevant info
+    return imagesToProcess.map((image, index) => ({
+      originalImageIndex: work.images.indexOf(image),
```
Comment (r1b) [inline server/images.js]: Be defensive and check images exists

File: server/images.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    return {
+      file: imageDataFilename(filename),
+      width,
+      height
+    }
+  }
+
+  const works = await Promise.all(data.works.map(async (work, index) => {
+    const processedWorkItems = processedWorks[index]
+    const images = await Promise.all(work.images.map(async (image, index) => {
```
Comment (r1b) [inline server/images.js]: Be defensive and check images exists

File: server/purify.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    updateProjectFile(data)
+  ))
+}
+
+async function processDesigners () {
+  const designers = await getDesigners()
+
+  const impureDesigners = designers.filter(designer =>
+    designer.slug !== slugify(designer.slug) ||
+    designer.works.filter(work =>
```
Comment (r1b) [inline server/purify.js]: Check if works exists

File: server/purify.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+      work.slug !== slugify(work.slug)
+    ).length > 0
+  )
+
+  const pureDesigners = await Promise.all(impureDesigners.map(designer =>
+    Object.assign({}, designer, {
+      slug: slugify(designer.slug),
+      works: designer.works.map(work => Object.assign({}, work, {
```
Comment (r1b) [inline server/purify.js]: Check if works exists

File: server/purify.js [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      }))
+    })
+  ))
+
+  await Promise.all(pureDesigners.map(data =>
+    updateDesignerFile(data)
+  ))
+}
+
+function slugify (str) {
```
Comment (r1b) [inline server/purify.js]: Do we need to reimplement this? Looks fine but there are probably other routines out there that handle unicode & urlencoding

**Comment by r1b** [comment]
References to images in the data need to be prefixed with `/public/static/` to show up in the CMS

File: src/layouts/Video.js [hunk 8]
```diff
@@ -62,6 +62,9 @@ const vimeoEmbed = videoId => (
 
 const Video = ({ video }) => {
   const { vimeoId, caption } = video
+  if (!vimeoId) {
+    return null
+  }
```
Comment (r1b) [inline src/layouts/Video.js:67]: lol nice

Comment (r1b) [reply src/layouts/Video.js:67]: as an aside, how do you feel about an `autoplay` option in the CMS?

**Review by r1b — APPROVED** [review]
Looks good just the data image path and defensive array checking are sufficient to merge

File: src/layouts/Video.js (hunk 8 above)
Comment (kevin-roark) [reply src/layouts/Video.js:67]: lol yep

**Pull request merged**